DRUP-530 Add swagger formatter for API Docs

arlina-espinoza opened the pull request:
Adds the swagger formatter to the default view of the API Docs. This PR:

- Adds the swagger_ui library.
- Adds the swagger_ui_formatter module and config.

Important note:
The _swagger_ui_formatter_ module uses drupal libraries, not composer, to load Swagger UI, and as such, it expects the library in _web/libraries_, and not in _vendor_. To allow for an automated installation I've added a composer "copy dir" plugin (_slowprog/composer-copy-file_), and we would need to update _apigee/devportal-kickstart-project-composer_ "scripts" and "extra" sections to the following:


```
"scripts": {
    "post-install-cmd": [
      "DrupalComposer\\DrupalScaffold\\Plugin::scaffold",
      "SlowProg\\CopyFile\\ScriptHandler::copy"
    ],
    "post-update-cmd": [
      "DrupalComposer\\DrupalScaffold\\Plugin::scaffold",
      "SlowProg\\CopyFile\\ScriptHandler::copy"
    ],
    "quick-start": [
      "composer install",
      "php web/core/scripts/drupal quick-start apigee_devportal_kickstart"
    ]
  },
  "extra": {
    "installer-paths": {
      "web/core": ["type:drupal-core"],
      "web/libraries/{$name}": ["type:drupal-library"],
      "web/modules/contrib/{$name}": ["type:drupal-module"],
      "web/profiles/contrib/{$name}": ["type:drupal-profile"],
      "web/themes/contrib/{$name}": ["type:drupal-theme"]
    },
    "enable-patching": true,
    "copy-file": {
      "vendor/swagger-api/swagger-ui/": "web/libraries/swagger_ui"
    }
  }
```
When testing this PR, be sure to update your project's root composer with the above.

arlina-espinoza committed:
DRUP-530 Added swagger ui formatter for API Docs.

arlina-espinoza committed:
DRUP-530 Added swagger ui formatter for API Docs config.

jacine commented:
@arlina-espinoza One of the questions I had, in looking at the `composer.json` included with this profile (not in this PR, just in general), is whether or not it makes sense, or even works at all, having this composer config setup with installer paths, etc. I wonder if we remove all of that and rely on [apigee/devportal-kickstart-project-composer](https://github.com/apigee/devportal-kickstart-project-composer) with an Asset Packagist setup, would that be acceptable (I'm really not sure)?  I think if we can do that, then we wouldn't need to copy it or do anything special to get it in the libraries directory.

arlina-espinoza commented:
@jacine The installer paths and other keys are not needed in this profile - composer will only read that configuration from the project's root composer file. However, I find it useful for documentation to keep it here.
Also, I'm not sure if a package like this exists in Asset Packagist, but in any case, it's the same issue: to add another repository like asset packagist requires editing the project's main composer file.
I've got a PR on the Kickstart Composer Project to add the necessary changes:
https://github.com/apigee/devportal-kickstart-project-composer/pull/3

jacine commented:
Thanks @arlina-espinoza. I was confusing [composer-installers-extender](https://github.com/oomphinc/composer-installers-extender) with asset packagist! Sorry! I think it can be done both ways.  Going to test this now with the other now. :)

arlina-espinoza committed:
DRUP-530 Use custom repo to add Swagger UI to web/libraries.

arlina-espinoza commented:
@jacine I've switched to use a more common drupal approach defining a custom package in the repository instead of the composer copy plugin, like we talked about. To test, add the following to the "repositories" on your project's composer root file:

```
"swagger-api/swagger-ui":     {
      "type": "package",
      "package": {
        "name": "swagger-api/swagger-ui",
        "version": "3.22.0",
        "type": "drupal-library",
        "dist": {
          "url": "https://github.com/swagger-api/swagger-ui/archive/v3.22.0.zip",
          "type": "zip"
        },
        "require": {
          "composer/installers": "^1.2.0"
        }
      }
    }
```

I've updated the corresponding PR on the Kickstart Project Composer to reflect this change: https://github.com/apigee/devportal-kickstart-project-composer/pull/3

minnur reviewed (approved):
LGTM lets get this merged
